src/cryptsetup/cryptsetup-tpm2.c

```diff
@@ -1,14 +1,63 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
 #include "alloc-util.h"
+#include "ask-password-api.h"
 #include "cryptsetup-tpm2.h"
+#include "env-util.h"
 #include "fileio.h"
 #include "hexdecoct.h"
 #include "json.h"
 #include "parse-util.h"
 #include "random-util.h"
 #include "tpm2-util.h"
 
+static int get_pin(usec_t until, AskPasswordFlags ask_password_flags, bool headless, char **ret_pin_str) {
+        _cleanup_free_ char *pin_str = NULL;
+        _cleanup_strv_free_erase_ char **pin = NULL;
+        int r;
+
+        assert(ret_pin_str);
+
+        r = getenv_steal_erase("PIN", &pin_str);
+        if (r < 0)
+                return log_error_errno(r, "Failed to acquire PIN from environment: %m");
+        if (!r) {
+                if (headless)
+                        return log_error_errno(
+                                        SYNTHETIC_ERRNO(ENOPKG),
+                                        "PIN querying disabled via 'headless' option. "
+                                        "Use the '$PIN' environment variable.");
+
+                pin = strv_free_erase(pin);
+                r = ask_password_auto(
+                        "Please enter TPM2 PIN:",
+                        "drive-harddisk",
+                        NULL,
+                        "tpm2-pin",
+                        "cryptsetup.tpm2-pin",
+                        until,
+                        ask_password_flags,
+                        &pin);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to ask for user pin: %m");
+                assert(strv_length(pin) == 1);
+
+                if (!tpm2_valid_pin(*pin)) {
+                        return log_error_errno(
+                                        SYNTHETIC_ERRNO(EPERM),
+                                        "Incorrect PIN length (must be at least 4 characters)!");
+                }
+
```

Comment:
it makes a ton of sense to validate PIN syntax when setting the PIN, but when checking it, I think we should just pass the value along as is. i.e. PIN policies might change over time and it should always be allowed to try PINs that might not be OK by today's standards but might have been by yesterday's standards.

(it might make sense to maybe as a friendly hint once the authentication failed with a PIN to check if it is superficially OK looking and then inform the user if it isn't, but that'd just be the cherry on top... I mean, we do no such thing for regular paswords...)